File: built-in plugins dsp/mb-synth-mono.cpp
/**
 * MB Mono Lead
 * Category : instrument
 * Type     : analog
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Fat monophonic lead
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SYNTH_MONO_H
#define MB_SYNTH_MONO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSynthMono : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-synth-mono";
    static constexpr const char* PLUGIN_NAME    = "MB Mono Lead";
    static constexpr const char* PLUGIN_TYPE    = "analog";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float glide = 0.1f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbSynthMono() = default;
    ~MbSynthMono() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.glide = std::clamp(params.glide, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Mono Lead
        return input;
    }
};

#endif // MB_SYNTH_MONO_H

File: built-in plugins dsp/mb-synth-mono.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSynthMonoPlugin: PluginDefinition = {
    id: 'mb-synth-mono', slug: 'mb-synth-mono', name: 'MB Mono Lead', category: 'instrument', type: 'analog', version: '1.0.0',
    description: 'Fat monophonic lead', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'sawtooth', detune: 0, gain: 0.6 }, { type: 'square', detune: 0, gain: 0.4 }],
    envelope: { attack: 0.005, decay: 0.2, sustain: 0.7, release: 0.3 },
    parameters: [
      { id: 'glide', name: 'Glide', type: 'float', defaultValue: 0.1, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { glide: 0.1, volume: 0.8 },
  };

export default MbSynthMonoPlugin;
